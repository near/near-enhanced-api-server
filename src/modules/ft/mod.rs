use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    // I'll drop these comments, leaving it only because it's easier for the review
    // app.service(
    //     web::resource("/accounts/{account_id}/NEAR")
    //         .route(web::get().to(resources::get_near_balance)),
    // )
    app.service(
        web::resource("/accounts/{account_id}/FT").route(web::get().to(resources::get_ft_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/FT/{contract_account_id}")
            .route(web::get().to(resources::get_ft_balance_by_contract)),
    )
    // .service(
    //     web::resource("/accounts/{account_id}/NEAR/history")
    //         .route(web::get().to(resources::get_near_history)),
    // )
    .service(
        web::resource("/accounts/{account_id}/FT/{contract_account_id}/history")
            .route(web::get().to(resources::get_ft_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_metadata)),
    );
}
