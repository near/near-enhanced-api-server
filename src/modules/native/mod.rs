use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app.service(
        web::resource("/accounts/{account_id}/balances/NEAR")
            .route(web::get().to(resources::get_near_balance)),
    )
    .service(
        web::resource("/accounts/{account_id}/balances/NEAR/history")
            .route(web::get().to(resources::get_near_history)),
    );
}
