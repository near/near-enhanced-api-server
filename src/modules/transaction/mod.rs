use paperclip::actix::web;

// mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app.service(
        web::resource("/transaction/{transaction_hash}")
            .route(web::get().to(resources::get_transaction_by_tx_hash)),
    )
    .service(
        web::resource("/transaction/receipt/{receipt_id}")
            .route(web::get().to(resources::get_transaction_by_receipt)),
    )
    .service(
        web::resource("/transactions/{account_id}")
            .route(web::get().to(resources::get_transactions_by_account)),
    )
    .service(
        web::resource("/transactions/{account_id}/{contract_id}")
            .route(web::get().to(resources::get_transactions_by_account_on_contract)),
    )
    .service(
        web::resource("/receipts/{transaction_hash}")
            .route(web::get().to(resources::get_receipts_by_tx_hash)),
    )
    .service(
        web::resource("/receipts/actions/{account_id}")
            .route(web::get().to(resources::get_action_receipts_by_account)),
    )
    .service(
        web::resource("/receipts/actions/{account_id}/{contract_id}")
            .route(web::get().to(resources::get_action_receipts_by_account_on_contract)),
    );
}
