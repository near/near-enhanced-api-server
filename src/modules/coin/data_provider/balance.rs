use crate::modules::coin;
use crate::{db_helpers, errors, rpc_helpers, types};
use std::str::FromStr;

// todo change to near_balance_events when all the data is collected
pub(crate) async fn get_near_balance(
    pool: &sqlx::Pool<sqlx::Postgres>,
    block: &db_helpers::Block,
    account_id: &near_primitives::types::AccountId,
) -> crate::Result<coin::schemas::NearBalanceResponse> {
    let balances =
        db_helpers::select_retry_or_panic::<super::models::AccountChangesBalance>(
            pool,
            r"
                WITH t AS (
                    SELECT affected_account_nonstaked_balance + affected_account_staked_balance balance
                    FROM account_changes
                    WHERE affected_account_id = $1 AND changed_in_block_timestamp <= $2::numeric(20, 0)
                    ORDER BY changed_in_block_timestamp DESC
                )
                SELECT * FROM t LIMIT 1
            ",
            &[account_id.to_string(), block.timestamp.to_string()],
        ).await?;

    match balances.first() {
        Some(balance) => Ok(coin::schemas::NearBalanceResponse {
            balance: types::numeric::to_u128(&balance.balance)?.into(),
            metadata: super::metadata::get_near_metadata(),
            block_timestamp_nanos: block.timestamp.into(),
            block_height: block.height.into(),
        }),
        None => Err(errors::ErrorKind::DBError(format!(
            "Could not find the data in account_changes table for account_id {}",
            account_id
        ))
        .into()),
    }
}

// TODO pagination
// todo support legacy contracts
pub(crate) async fn get_coin_balances(
    pool: &sqlx::Pool<sqlx::Postgres>,
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    block: &db_helpers::Block,
    account_id: &near_primitives::types::AccountId,
    limit: u32,
) -> crate::Result<Vec<coin::schemas::Coin>> {
    let query = r"
         SELECT DISTINCT emitted_by_contract_account_id account_id
         FROM assets__fungible_token_events
         WHERE (token_old_owner_account_id = $1 OR token_new_owner_account_id = $1)
             AND emitted_at_block_timestamp <= $2::numeric(20, 0)
         ORDER BY emitted_by_contract_account_id
         LIMIT $3::numeric(20, 0)
     ";
    let contracts = db_helpers::select_retry_or_panic::<db_helpers::AccountId>(
        pool,
        query,
        &[
            account_id.to_string(),
            block.timestamp.to_string(),
            limit.to_string(),
        ],
    )
    .await?;
    let mut balances: Vec<coin::schemas::Coin> = vec![];
    for contract in contracts {
        if let Ok(contract_id) = near_primitives::types::AccountId::from_str(&contract.account_id) {
            balances.append(
                &mut get_coin_balances_by_contract(rpc_client, block, &contract_id, account_id)
                    .await?,
            );
        }
    }
    Ok(balances)
}

// TODO PHASE 2 change RPC call to DB call by adding absolute amount values to assets__fungible_token_events
// TODO PHASE 2 add metadata tables to the DB, with periodic autoupdate
pub(crate) async fn get_coin_balances_by_contract(
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    block: &db_helpers::Block,
    contract_id: &near_primitives::types::AccountId,
    account_id: &near_primitives::types::AccountId,
) -> crate::Result<Vec<coin::schemas::Coin>> {
    let (balance, metadata) = (
        get_ft_balance_by_contract(
            rpc_client,
            contract_id.clone(),
            account_id.clone(),
            block.height,
        )
        .await?,
        match super::metadata::get_ft_contract_metadata(
            rpc_client,
            contract_id.clone(),
            block.height,
        )
        .await
        {
            Ok(metadata) => metadata.into(),
            Err(_) => super::metadata::get_default_metadata(),
        },
    );

    Ok(vec![coin::schemas::Coin {
        standard: "nep141".to_string(),
        contract_account_id: Some(contract_id.clone().into()),
        balance: balance.into(),
        metadata,
    }])
}

pub(crate) async fn get_ft_balance_by_contract(
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    contract_id: near_primitives::types::AccountId,
    account_id: near_primitives::types::AccountId,
    block_height: u64,
) -> crate::Result<u128> {
    let request = rpc_helpers::get_function_call_request(
        block_height,
        contract_id.clone(),
        "ft_balance_of",
        serde_json::json!({ "account_id": account_id }),
    );
    let response =
        rpc_helpers::wrapped_call(rpc_client, request, block_height, &contract_id).await?;
    Ok(serde_json::from_slice::<types::U128>(&response.result)?.0)
}

impl From<coin::schemas::NearBalanceResponse> for coin::schemas::Coin {
    fn from(near_coin: coin::schemas::NearBalanceResponse) -> Self {
        coin::schemas::Coin {
            standard: "nearprotocol".to_string(),
            balance: near_coin.balance,
            contract_account_id: None,
            metadata: near_coin.metadata,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::modules::tests::*;
    use std::str::FromStr;

    #[tokio::test]
    async fn test_near_balance() {
        let pool = init_db().await;
        let block = get_block();
        let account = near_primitives::types::AccountId::from_str("tomato.near").unwrap();
        let balance = get_near_balance(&pool, &block, &account).await;
        insta::assert_debug_snapshot!(balance);
    }

    #[tokio::test]
    async fn test_coin_balances() {
        let pool = init_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let account = near_primitives::types::AccountId::from_str("patagonita.near").unwrap();
        let balance = get_coin_balances(&pool, &rpc_client, &block, &account, 10).await;
        insta::assert_debug_snapshot!(balance);
    }
    // todo add pagination tests when we finalise how it should look like

    #[tokio::test]
    async fn test_coin_balances_empty() {
        let pool = init_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let account = near_primitives::types::AccountId::from_str("olga.near").unwrap();
        let balance = get_coin_balances(&pool, &rpc_client, &block, &account, 10)
            .await
            .unwrap();
        assert!(balance.is_empty());
    }

    #[tokio::test]
    async fn test_coin_balances_by_contract() {
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("nexp.near").unwrap();
        let account = near_primitives::types::AccountId::from_str("patagonita.near").unwrap();

        let balance = get_coin_balances_by_contract(&rpc_client, &block, &contract, &account).await;
        insta::assert_debug_snapshot!(balance);
    }

    #[tokio::test]
    async fn test_coin_balances_by_contract_no_contract_deployed() {
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("olga.near").unwrap();
        let account = near_primitives::types::AccountId::from_str("patagonita.near").unwrap();

        let balance = get_coin_balances_by_contract(&rpc_client, &block, &contract, &account).await;
        insta::assert_debug_snapshot!(balance);
    }

    #[tokio::test]
    async fn test_coin_balances_by_contract_other_contract_deployed() {
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("comic.paras.near").unwrap();
        let account = near_primitives::types::AccountId::from_str("patagonita.near").unwrap();

        let balance = get_coin_balances_by_contract(&rpc_client, &block, &contract, &account).await;
        insta::assert_debug_snapshot!(balance);
    }
}
