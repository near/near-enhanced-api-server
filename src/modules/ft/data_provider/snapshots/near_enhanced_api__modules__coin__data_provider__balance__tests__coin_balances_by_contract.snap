---
source: src/modules/coin/data_provider/balance.rs
expression: balance
---
Ok(
    [
        Coin {
            standard: "nep141",
            balance: U128(
                6103689,
            ),
            contract_account_id: Some(
                AccountId(
                    "nexp.near",
                ),
            ),
            metadata: CoinMetadata {
                name: "NEXP",
                symbol: "NEXP",
                icon: Some(
                    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAQAAAAEACAYAAABccqhmAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMy1jMDExIDY2LjE0NTY2MSwgMjAxMi8wMi8wNi0xNDo1NjoyNyAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDpCN0Y4OTQ5NEMxQTJFQzExOTYyRDlEQzhBQzkzREZEOSIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDpCMEEzMTExNEEyQ0ExMUVDODFGN0YzNjM5NzNDOTI5MSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDpCMEEzMTExM0EyQ0ExMUVDODFGN0YzNjM5NzNDOTI5MSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M2IChXaW5kb3dzKSI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkI3Rjg5NDk0QzFBMkVDMTE5NjJEOURDOEFDOTNERkQ5IiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOkI3Rjg5NDk0QzFBMkVDMTE5NjJEOURDOEFDOTNERkQ5Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+ZGswBAAAGlFJREFUeNrsnQl4VFWWx08qe2JIAmGTNezLoIKKSyu4fqKj3Y1ru7SCrY6tPeq0COg3TSOtg+mRnlbHZVxwAVFcaLVttdUWUMFG3JCdIKssCVtCyJ6QOafeDYRQqdSrvOW+9/6/7/t/IEKq6tx7/3WXc89LGLP4KgKeJo2Vr9SZ1YGVp9ShieTvZbGS1O/Tm/2cSlYVq45Vpn6/l7WHtbuJ5L+LWBuVqtAE3iUJIfAEMlgHsQazhqrB3lupq4Wv0WgKHU38ux2sTUpiCCtZq1lrlKkAGAAwQSbrRNbJ6tcTWANYiZq+365KpzX783rWOtZ3rG9YX7K+ZpWjiWEA4DAybT+TNUr9+i8+aZdENWMRXa3+TJYXK1ifsT5VvxahC7hHAvYAHCeDdRbrPNb5asAHGVkyfKS0gFWBLgID8BvdWJconUPGJhw4GtlQ/IT1V6VtCAkMwKt0YV3KkgCfwQohJKY4yPqcNZc1j7UTIYEB6E571hVq0I8ifTfuvEa92jMQM3idjONJAAPQI4ZqsN/MugzTe0eWCW+ynlGm0ICQxA+mpfHTiTWRjPPuBaxrMfgdIU3FeoGK/UTVFgAG4AhyLv8iayurgIwzeuAOA1QbSFvMUm0DYAC2TPMvIOOo6lvW9awUhEUbpC2uU23zkWqrBIQFBtBWktRgX8b6gIyze6A356m2WqbaDsluMIC4Bv44MnLaZbo/DCHxHMNU261RbQkjgAHEFI9fsFaxnmf1Q0g8T1/VlqtU26LPwwAiMoaMyyqvsPojHL6jv2rbr1VbAxhAmCGs95Wwi+x/TmjS3kNgAMElh/UoGZtF+EYI5oxP2v4x1RdgAAFBjoduYBWy/p2wORRkpO1/o/rCDRTAo8OgGUAfMs6JXyCjZBYApPqC9ImPydg0hAH40OknkFGM4lz0d9ACclV7OeueoMwMg2AAsumzhPXfdHQhTACaI33kj6rPDIcBePuzTSajFt0I9GtgkhHKBCb7eZz49YP1JKOyzHRWMvoyiJNk1Yc+UX0KBuABriSjEu1o9F9gEdKXlqm+BQPQFCm2OZOMqjG56LPAYnJU35qp+hoMQCMGqPXaePRTYDPjVV8bCAPQg7GspYTy2sA5pK99qfoeDMDF934/GfXh2qFPAodpp/reNC+PI6++cXl81husKYTKL8A9pO/9TvXFTBiAM0i9/U/8MP0CvmGs6pNdYAD2Itc3F7NGos8BzRip+uYQGIA9nE3Gk2Ly0deApuSrPno2DMBapJSTFHrE+T7QnVzVV6+GAVjDrazZhDLcwDukqD57KwygbUxiPUl4xh7wHiHVdyfBAOJDjlceQj8CHuch1ZdhACaYpgSAH5C+fD8MIDam6OyYALShX0+BAUTnHl2dEgALkL49EQYQmV+TUYoJAD9ToPo6DKAJ15BRnx2AIPCY6vOuo0PlU6nEKs9uw1GfSfYv2EX7Zu1w/lsjO5G6TO5HyZ1S0Qjxkaj6/E4y7hAEdgYgT3CdR0jyMcXB6oO0981trgz+8OuX1tP2e9dS+felaIz4SVF9f1hQDaA76z1WNvpC7NSX1lLRYxuo7L09rr+X3Y9sDhsRiJtsNQa6B80ApKba39z84F6ktriatt+/jmpWV2jznsSIip/aGJ6VgDZ9EWYExQCkiMIc1nFo+9iR6bZMu2X6rRuVS8tox/TCsEGBuJfCc8iF4jaJ/X7leCk9SYa4FW0eOzLNLpm9Q+v3eHB/PZV/VUJJXVMppXMaGs08g1gNrIV+NgB5JPP/UbAfSx77oOJp9a5nN1H5whJPvN+G6gaqWFJKoewQpfbORAOaZxQZBW7X+3EJ0EdNc3DcF+N6X6bVMr32GnI6sXvWFuwLxDceX1ZjxVcGIBscUkEVBT1iWVOvK6OdD62nuq3eXVOXLygJn1bIqQUwRa4aKxl+MoD/JeMpvaAVJLmnuGCjlpt9ZpHTCjm1qN5SgYY1h4yVx/1iANcSntgT03rfzeQe2z4XG9nO+9cjacg849TYsZWEMYuvsvPn92V9y8pCe7aMTJOLn9ms1fm+HWRd1IHaX9YNDR47sgF0IqvQizMA2eybjcEfHR2Te2zrze/toR0Pr8fmoAnPZM0iG+/s2GkA97FORRu2jM7JPXbuC2y7dzWShmLnFDWWPLUEGE7GE1ST0X6RkfW+Dvn8biE3CtuP606Zx+EqSCwTRfVl+o0XZgAyXXkWgz8yMv2V3PkgD/5wHHjWI5eJSt7fiU7ROslqTFm+FLAjE/Bu1vVos8jr/aL/2RCI9X6sVK8qp7r91ZQ2KIsSkvCc1yh0ZUnHWaTzEkB2/Zez0tFeRyLJPbuf2hKo9b4ZUgZnUN71PVBkpJVuRMYlOstSha1cAoh9P43BfzR+Su6xC5kVSfYjkoaikq7GWIKOBjCOjPJeoMl634/JPXbuC0jSUNmSvQhGy8iDRy1LrLNqCZDDWsfqiPYx8FJyT0JqQvgmn05I0lDOxV0plIqLoxEoZvWXyWVbf5BVm4AF+PY/TPgm34OFVPdjjd5vNJRAHa7vTjmXd6XKdQfCd/q1WRIUVlLVhgOUeWIONgePRu5aS03BD3WYAQxlfUd6VBh2HUnukeMt/b/1Q9Tx1l6UflzWoeWK3N7TbcaCCsQtIplUUkmoTWnCVswAXmINQHt4o3JPuNGzk6jzb/tQ2sDDRTvkWzbr9PZ0sK4u/O2rC7I0KfvHHkrOT0OloSORL9xerFfdNIALWFOD3hJeqtyT3DmVOt/Tl5K7RR5M6UPahQebVPbRifD7SWmgtP7HYOgfZiDrC9YPbhiALMxeIyNBIdDrfa8k96T2y6DOE/pSYm70JE35ps08LSdc40+nzUFJGqrZUUHpw7KxL3AYGcBPu2EAl7HuDHLkJbmn+M8bqb5Y/6o3GSOyqdNvelMoPbaKbImZSXTMqblUtaWC6nfr8/nqtteENwfT+mWG3yMIfwFL8t1qJw1AzmbmsjoFNeqS3LPnia3aHZ9FIuvcPMob352/Nc0dqYXSEsO78A0N9VrtC4ghyewkJT+dkjtgc5CMjfinyKgq7IgB/II0esKp0+v9fW9vp9I3i/V/szxLzr2iK+WM7cK/j2/KLFNt2ReQSr9V3x/Q5qOJ8ZYvKkEFYgP5Il7LWuGEASSptX9e0KIcTu55chNVfrFf/7HPAzfv5p6UNbqDNfsHPMhSB2dQ5YoyrWY9YkpycpHa95ig7wsMU7OAg3YbgDzW+OagRdczyT0ydc9IpE539qaM49tZ+nNlup0xPFvbpKGMoVnhZUtAEaeXnIDvTX1RmEwESlDTjCFBiqxXknvC07P2ydTprvwWj/msWgbtfn6zds8sQNJQeGweZ2YvwOwM4KesO4IUUa8k9wgp3dPCZ/xJNg8AmWpnnpQbPpeXozmd9gUCnjQkewGm8gLMGoBUJekZhEh67bFcaUOOoU7/kU+JWc4djUlSjq5JQ7IvIJuXATWB2XYsAaTO3zdBiKCs94uf2OSZJ/NknpZLHcZ1d20TTOIld/l1q3eQfnIW5Y3vFcQbhSPIKMdv6QxAbvwd7/fIeSm5R8i+qBO1v7YbJSS6twMuCTlZZ+WFN+J0Sxqq+L6UZwJZQUsaksIhb1k5A5BpxVYyriD6Fknu8UzxjlACD/xjeeB10OYtybKp5N0d2hU8DWAF4hq1VC+yagZwOxmP9vbtel/L5J4WvtTlKm+n23pR5ql6PWu1MWkosUsyVX6tT65EAB9bLmehuymGAqKxGIB0w5lknDP6Dm2TeyTqEQ5zZJNPNvvSBut7Ky61e7q2SUMBqkAsM4DHrTCA0WSU+vYdXkruEZI6pVCXiX3Dx32605g0VL29Uqt9gdpNVUFJGpIv7PmszW01gAfJSC7wFZLcU/TgD564zBP+Vu0jV3n7hBN9PDMPzUwKXyaqLa4Mb8hpM+tjQzrwz33hWVRStq+fXyN7dvOiTjRb2QRsz9rG8lVWhY6P5ZIpaUNdZDNKP6Eddfy3npSQ4t3jLHkCUOkb+l2gyruzl583B6tY8jjmvfHOAMaxfuaXaOia3JOQFqKGmsiDX3b5837Vw/RVXt1oTBqqWnNAq1mXz5OGktQS4KuW/kJrveoaX633pxdql78uO/oNVREucCUQ5VzWhdr/slv4yM8PyDet5OrLsZxO+Pyx5dfEuwSQgoMbycKnkLiF1x7LJcsByeyTDD8/ggrEjiLTrXxqYTMw2hLgJjKKfnoaL1XuCXfC9ETqdEfv8A66X9G5ArFUGkrqmuqny0TyBS77eIvNGsAMVg8vf8t4pnJPY2PkJod3+lP7ZlAQ0DlpyGcViGWD4zkzSwDZOdzq1em/lx7L1Yjc35d7/F465rMKeSCo3L/QbYmWeVYOtb+yux8uE8n0VxKDfmz+P1q6IXGJVwe/rjfTsrOzKSsrixISEo7Q/v37KblXGqXcmBuu5BNEUntmhNfeu1/aqpVply8oodqiGj88tjxBjeknY50B/IX186B0wNt7jqeLu1m33VFbWxse3OEAq18bGhrCahz4h6yZ/6zs4AG6fdkk2ldXSkFGlm21RVXh30/udwf1zOxuy+ssLF5Ec3e8HfPfl2ShRO8nDMntwLGxzADkk55HIP7pfLK5zpJLOfTw0Kk0YeXUQJuATLVlNiD0GdCH+mX1seV11mdtptTkjKCFV8a0ZAYekZIZaXFzBgvPX3KYYzO60rRBkxAIYBcypn9ylOlG+IsXIlbuIN94U/tPQCCAXVwUiwGMQZzc45S8k2hi/m0IBLCDC1ozAKn8Mwxxcpezu4yiKzpfjEAAq5Gx3TmaAYxGjPTgxj7XwQSAHYyOZgBnIj76cE2vy+n4zMEIBLCSUdEM4HTERx/SEtNo6tBJMAFgJWe0ZAByMHo84qOfCUwadCflJmUjGMAK5PLPMZEM4CRqOTUYuEhuipEoBBMAFiD55iMiGcBIxEZfkCgELGRkJAMYjrjoDRKFgEWMgAF4FCQKAQsY3twA5FliAxAXb4BEIdBG+qsxf8gAhqjNAeARkCgE2oCM9UFNDQAHzR4EiUKgDQxpPgMAHgOJQqANDG1qAH0QD++aABKFQBz0aWoA/REP74JEIRAHA5oaQF/Ew9sgUQiYpG+jAeSw8NXhA5AoBEwgzwrIEQPogVj4ByQKARP0gAH4ECQKATMG0BlxsI/5Oz+l7RU7HH9dSRS6sMPZaAAQjS5iAB0QB/sor6+kB9b+iarqqxx/7Vv63oAcARCN9mIAHREHe9lYtZWmrixw3ASQKARaoSNmAA6xrHw1zVj7uOOvKyZwR79bkCMAItEBBuAgn5cupZkbZjv+upIjgEQh0JIB5CEOzvF60buumcC9/e5AA4CjDCAHcXDeBJbs/srx1x2WOxSJQqAp4USgdoiD80wtfNgVE5BEIXkcOgBMthhAGuLgngm4kSNwcbcLkCgEhFQxgFTEwT0mrJyKRCHgFmliAHgWgIvsqysNmwAShYALJIkBZCIO7psAEoWAC2SGEAM9kEQhMQHH54BIFAo0MADNTACJQgAGEGCQKAScNoAKhEE/E0CiEHCAcjGAWsRBP5AoBBygTgygGnHQ1wSQKARspEoMoApx0BckCgEbqRYD2I846AsShYCNlIoBlCEO+psAEoWADRwQAyhBHPQHiULABkpgAB4zASQKASsnl2IAxYiDd0CiELCQXWIAuxAH75kAEoWAVQawE3HwHkgUAhawUwxgB+LgXRNYX7bB8ddFopC/DGAr4uBdpqwpQKIQiJetMACP05gotK/G+cMcJAp5ni2Nx4CliIW3TaBgzSNIFAJmkAzgksZ6AD8gHt6mMVHIDRNAopAnCY/5RgMoRDz8YQJzNr/h+OsiUciTFDY1gA2Ihz9AohCIZwawCvHwlwnM3/mp46+LRCFPsbKpAaxGPPzFHzc+gUQhEI1VzWcA9YiJv0CiEGgBGetrmhpAJWsd4uI/3EwUOiP7ZDSAnhSqMX9EWfBvERf/4Wai0N0Db0eOgJ4cGuswgICYgJuJQvlpPdAIevFNJAP4EnHxL24mCv3nwN8iR0AvlkYyANkyrkNs/G0CSBQKPLIB+HUkA5AnBC1DfPyNm4lCd+bfjAZwnxWsA5EMQPgC8QmGCbiRKCQ5AkgUcp3Pm/5HcwP4FPEJBm4mCo079ko0gHt8Gs0AFiI+wcGtRKGrel2KRCH3WBjNAKRC8HLEKDggUShQyNguimYAwgeIU3BAolCg+HvzP4hkAO8jTsEzASQKBYL3YjGARdTkmAAEAyQK+R4Z04tjMYAa1seIVzBN4OkfXnT8dZEo5AgypqtjMYCIawUQDN7fMx+JQv7kw0h/2JIBvMtqQMyCiSQKvbvN+e8AJArZhozlv5oxgB9ZSxC34PL4lueRKOQflqgxHbMBCG8gbsFGEoWW71vp+OtKolD3jGPRANbxZkv/ozUDwDIg4Exf/6griUJyOgAsm/6/Ho8BbCbjSBAEmMZEITdMAFjCIjWWTRuAMAfxA2ICj65/2vEcAWAJUcdwawYwl4VWB64lCoE2UaXGcNwGsJc1D3EEjSbgRqIQiJt5agzHbQDC04gjaMStRCEQF8+09hdiMQApIIBnBoBDuJUoBEwhY3ahFQYgxwjPIZ6gKW4lCoGYmUkxHOOHYvxhL5BxSQiAQ7iVKARapUaNWbLKAKRS0KuIK2iOW4lCICoyVousNADhMcQVNAeJQlry51j/ohkDkAUfyoaDiCaARCFt+IhMPOYvZPKHFyC+IBJIFNKGGWb+slkDeIe1CjEGLZkAEoVcRXZkP7TTABowCwDRQKKQqxSQyRu8oTheRC4X/IBYg5ZAopAryBNeXjH7j+IxAHmC8AOIN4gGEoUc5wGK4+neoThfTOZ46xFzEA1JFIIJOILMyGfF8w/jNQBxmt8j7qA1Htn4DHIE7GdKPN/+bTEAQbKNliH2IBpIFLKdZdSGLN22GMBB1iTEH8RiAg+s/RNyBOxhohqLjhuAIFu9eJgoaJWNVVuRKGQ9Uuv/w7b8gJAFb+KeeNcfIGBz1fLVNGPt4wiENciNv7vb+kOsMIAVrCfQHiAWPi9dikQha5DLeYU6GIAgJwK70CYgFiRRaO5mlJpsAzLWplnxg6wygBLWZLQLiJUXtr+GHIH4kbG2XycDEJ5nfYa2AbGCRKG4mK/GGulmAHIJ4VZC6TBgAiQKmaKSdQtZ+Mi+kMVvUK4KP4R2ArGCRCFT3E8Wp+CHbHiTD7KWo62AGRNAolCrSJWfGVb/UDsMQJYA41m1aDMQK0gUiork2dxENuTbJIxZfJVdb3qKmrIAANrGVLvGUsjGN/1frCVoOwDaxJdqWU1eMwCZrlzHKkMbAhAXZWoM1XnRAATZsfw12hGAuLidLEj3ddMAhJdZKBULgDleojir/OhmAMJtrO/QpgDExDKnZs5OGUAF6zLWPrQtAFGRMXKpGjO+MQBByhZfS22oXgKAzzmoxsgGp14w5PAHfJ8susYIgA+ZpsYI+dUAGj/kO2hrAI7gHTe+HN0wALnJdDUhSQiARpaoMdEQBAMQZIPjEtaPaHsQcGQM/Iwc2vTTxQAEKWt0EasUfQAElFI1BorcegMhlwMg14blyANFREDQqFF939Wr8yENAvEJ6wZWPfoECAj1qs9/4vYbCWkSEHm0kaWljgDQlAbV11/V4c2ENArMTNZd6B/A59yl+jrBAI7mUdZ96CPAp9yn+jjBAFpmOusP6CvAZ/xB9W2CAbSOlBP7PfoM8Am/V32aYACxI2mR96LvAB9M+7W9/xLSPHjyjAGpJYAbhMBrHFR9d7rObzLkgUA+yfolIVkIeIda1Wef1P2NhjwS0DlkpEyWoG8BzZFCnv+q+izBAKzjH6wzWJvQx4Cm7GSNZn3klTcc8liAV7JOYy1FXwOaITUvTybjEV4EA7DXZc9mvYU+BzRBinmcSR683h7yaMDLySgy+gDh/gBwjwbVB8eyDnjxA4Q8HHw5Zvkd63KvBh94GulzV6g+6Nlj6pAPGmKeWnutQJ8EDrFC9bk3vf5BQj5pkDWsUwhPIAL2M0v1tTV++DAhHzWM1FQbR0ZxReQLAKuR8/3rlSr88qFCPmwoKbRwPGsh+iywCKnaeyI58Kw+GIA1bGGdQ8ZFjFr0XxAn8lhuucknCWiFfvyAIR83nuzMykWMU8ljyRlAC75Ta/1pyggIBuBNvmGNZE1mVaJfg1aoVH3lZNV3fE0oII0qDl7AOo41H30ctMB81UcK/PytH0QDaGQ961zWTazd6O9AsVv1iXNVHwkMoQA2tqRvPsfqz5pBqDMQZGpUH+iv+kTg0spDAW58yRWYwBrK+gvGQuB4S7X9BApw3kgI/SA85ZNHNF3IWoZw+J5lqq3HBm26DwOIzgesEaxr0DF8a/TXqjb+AOGAAURCcgdeYQ1m3cjagJB4no2sm1WbziEUmIUBxIAcAT3PGkjG/YJVCInnWK7abgDrWQrIsR4MwHojkBuGw1g/Zy1CSLTnY9YYMu6DvIiBDwOwamnwNhk54SPVVBJ3DPRB2mI2azjrfNbfCZWiYAA2IQVJZTOpBxkpo9gwdI/1qg2kLaQO/3cICQzAKYrISBmVNabcPHyZVY2w2E61moGdo2JfoNoCxEHCmMVXIQrW0Z4lAZVacaNYiQiJJdSzPmO9xprL2ouQwAB0pwsZBUslwKdjtmUa2XdZrAb8G2SUgwcwAE/SjfVT1iVkPNMgDSGJSBUZN/LeJWPTdRtCAgPwGxnKBM5XGhLweEiOxUdKMvgr0EWcIwkhcBzp4H9TEjqxfqIkjz2TQhTJPv3sclwnpyhfkJFTISpGl8AMABwmnXUSGaXM5Fz7BDJ2u722oSgbd+vIOJqTkmz/ZH1FqMqEGQCIigyQz5SamsIgMq6vSk57Pqu3UleX3+8OMp7YLJK8+9VqWr8agx0GAKwzhW8pcnHT9CaG0JmVx+rI6tBMqaxsMk4j2jWbUciypDGHQe7GN6j/3tNMu8ionlPUZMBjkHuY/xdgAEYrm1da1anaAAAAAElFTkSuQmCC",
                ),
                decimals: 4,
            },
        },
    ],
)
